ouroboros-consensus/ouroboros-consensus-test-infra/src/Test/Util/HardFork/OracularClock.hs

```diff
@@ -0,0 +1,242 @@
+{-# LANGUAGE Rank2Types          #-}
+{-# LANGUAGE RecordWildCards     #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+-- | Intended for qualified import
+--
+-- > import Test.Util.OracularClock (OracularClock(..))
+-- > import qualified Test.Util.OracularClock as OracularClock
+module Test.Util.HardFork.OracularClock (
+    OracularClock(..)
+  , new
+  , onSlotChange
+  , withinEachSlot
+  , withinTheCurrentSlot
+  ) where
+
+import           Control.Monad (replicateM_, void)
+import           Data.Foldable (toList)
+import           Data.Time
+import           GHC.Stack
+
+import           Ouroboros.Consensus.Util.IOLike
+import           Ouroboros.Consensus.Util.ResourceRegistry
+
+import           Cardano.Slotting.Slot
+
+import qualified Ouroboros.Consensus.BlockchainTime as BTime
+import           Ouroboros.Consensus.Util.STM (onEachChange)
+import           Ouroboros.Consensus.Util.Time (nominalDelay)
+
+import           Test.Util.HardFork.Future (Future, futureSlotLengths,
+                     futureTimeToSlot)
+import           Test.Util.Slots (NumSlots (..))
+import           Test.Util.Stream
+
+-- | A clock that knows the future
+--
+-- This clock's closure contains a 'BTime.SystemTime', a 'Future', and a
+-- 'NumSlots'. As the system time advances, the clock maintains a notion of the
+-- current 'SlotNo'. Once all 'NumSlots' have passed, the clock is /exhausted/
+-- and most of its methods block indefinitely from then on.
```

Comment:
This is indefinite blocking behaviour important? Could we make it throw an exception instead? "Blocked indefinitely"  bugs are notoriously difficult to debug.

Reply:
I will try the exception and see if that's as workable.

Note that there's crucially one method that does not block indefinitely (`waitUntilDone` instead blocks **until** all other methods will start blocking) and `Network.hs` uses that one to kill the entire test. So at least `Network` is pretty robust here.

Reply:
It was difficult to rely only on an exception and to "expect" that exception in Just The Right places.

Instead, I replaced the indefinite blocking with a small delay (arbitrarily: two slots), which is followed by an exception that is indicative of a bug: our clean-up (especially in the `SimM`) should be more prompt than that. (Tests currently pass because it currently is prompt enough. In fact, I anticipate that any non-zero delay suffices, so I'll try tightening it.)

Reply:
I moved it down to 1 picosecond, the minimal delay. Looks good.